backend/routes/items.js

```diff
@@ -1,73 +1,90 @@
-      console.error(err);
-    }
-    res.send(items, 200);
-  });
-});
+router.get("/getAllItems", async (req, res, next) => {
+  const items = (await Item.find({}).sort({name: 1})).map(x => x.toObject());
+  const batches = (await Batch.find({}).sort({itemName: 1, outDate: 1})).map(x => x.toObject());
 
-router.post("/addItem", (req, res, next) => {
-  // validate body params
-  const itemName = req.body["itemName"];
-  const itemId = req.body["itemId"];
-  const weight = req.body["weight"];
-  const today = toUTCMidnight(new Date());
+  let batchIndex = 0;
 
-  if (itemName === undefined && itemId === undefined) {
-    res.send("Either itemName or itemId is required!", 400);
-    return;
+  for (const item of items) {
+    // Skip batches corresponding to items with names before this item's name.
+    // This loop should only run if there are batches corresponding to items that
+    // no longer exist.
+    while (batchIndex < batches.length && batches[batchIndex].itemName < item.name) {
+      console.warn("Batch does not correspond to any existing item: ", batches[batchIndex]);
+      batchIndex++;
+    }
+
+    // Add all batches of this item to an array.
+    item.batches = [];
+    while (batchIndex < batches.length && batches[batchIndex].itemName === item.name) {
+      item.batches.push(batches[batchIndex]);
+      batchIndex++;
+    }
   }
 
-  if (weight === undefined) {
-    res.send("Weight is required!", 400);
+  res.status(200).send(items);
+});
+
+router.post("/addItem", async (req, res, next) => {
+  // Validate request body fields.
+  try {
+    Assertions.assertObject(req.body, {
+      itemName: Assertions.assertString,
+      itemId: Assertions.assertString,
+      weight: Assertions.assertNumber,
+      outDate: Assertions.assertDateString,
+    });
+  } catch (e) {
+    res.send("body" + e.message, 400);
     return;
   }
 
-  // create batch object
-  const batch = new Batch({
-    date: today,
-    pounds: weight,
-    fulfilled: false,
-  });
+  const { itemName, itemId, weight } = req.body;
 
-  // check if the item already exist
-  Item.findOne({ name: itemName }, (err, item) => {
-    if (err) {
-      console.error(err);
-    }
+  const outDate = toUTCMidnight(new Date(req.body.outDate));
 
-    if (item === null) {
-      // item does not exist, so create one and add the batch
-      let newItem = new Item({
-        name: itemName,
-        itemId: itemId,
-        batches: [batch],
-      });
-      newItem.save();
-    } else {
-      // if item already exist with same date, combine the two batches
-      for (let batch of item.batches) {
-        if (batch.date.toString() === today.toString()) {
-          // combine batches
-          batch.pounds += weight;
-          item.save();
-          return;
-        }
-      }
+  // Use today's date as the in date, unless the out date is in the past.
+  // This ensures that the out date >= the in date.
+  let inDate = toUTCMidnight();
+  if (inDate >= outDate) {
```

Comment:
Should we throw an error if the outDate is in the past? I'm not sure it makes sense to arbitrarily set the inDate as the outDate just because the outDate is in the past.

Reply:
I agree with you. I'll change the code to return an error 400 if `outDate` is in the past. Thanks for the suggestion!